llvm/lib/Transforms/Vectorize/VPlan.cpp

```diff
@@ -950,6 +957,26 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
   }
 }
 
+bool VPlan::isExitBlock(VPBlockBase *VPBB) {
+  if (!isa<VPIRBasicBlock>(VPBB) || VPBB->getNumSuccessors() ||
+      VPBB == getScalarHeader())
+    return false;
+
+  VPRegionBlock *RegionBlock = getVectorLoopRegion();
+  if (!RegionBlock)
+    return false;
+
+  // The block must be a successor of the region block.
+  if (llvm::is_contained(
+          vp_depth_first_shallow(RegionBlock->getSingleSuccessor()), VPBB)) {
```

Comment:
I am not sure I completely follow the logic here. We return true if `VPBB` is any block after the middle block, basically relying on the checks above to filter out all non-VPIRBBs or VPIRBBs with successors.

Simpler to `return isa<VPIRBasicBlock>(VPBB) && VPBB->getNumSuccessors() == 0;`

Reply:
OK, that's fair enough. I wasn't sure if those two checks would always be enough, but it makes sense.